Add the possibility to compute the background model as average of model for sub runs

mdebony opened the pull request:
This PR has the possibility to compute background model as the average model across model created for different part of the run. Could improve the accuracy of the model when background is changing rapidly.

This is compatible with zenith binning and interpolation.

gabemery commented on acceptance_modelisation/toolbox.py:
```diff
@@ -82,3 +84,68 @@ def get_unique_wobble_pointings(observations: Observations, max_angular_separati
+
+    # Reshape the scale factor in order to be able to multiply it to the data cube
+    shape_data_cube = list(data_cube.shape)
+    for i in range(len(shape_data_cube)):
+        if i != 0:
+            shape_data_cube[i] = 1
+    shape_data_cube = tuple(shape_data_cube)
+    scale_factor_per_bin_reshaped = scale_factor_per_bin.reshape(shape_data_cube)
```
Here you have a scale factor of dimension 1 and len N, multiplying data of shape (N, ?, ?, ...) 
It should be possible without reshaping.

Also if you have a single scaling factor it is even easier

mdebony replied:
I've got an error during my testing, and I just added this reshape to be sure of what will be done. As the supplementary dimension are of order one, it should not involve any copy and thus slow down the code.

gabemery commented on acceptance_modelisation/toolbox.py:
```diff
@@ -82,3 +84,68 @@ def get_unique_wobble_pointings(observations: Observations, max_angular_separati
+    scale_factor_per_bin_reshaped = scale_factor_per_bin.reshape(shape_data_cube)
+
+    # Compute the weighted average of all mini irfs
+    data_cube_final = np.sum(data_cube * scale_factor_per_bin_reshaped, axis=0)
```
scale_factor_per_bin_reshaped could be out of the sum if single scaling factor.

gabemery reviewed (commented):
Docstrings need improvements.

observation_time as an array increase complexity for no apparent reasons

mdebony commented on acceptance_modelisation/toolbox.py:
```diff
@@ -82,3 +84,68 @@ def get_unique_wobble_pointings(observations: Observations, max_angular_separati
+        evaluation_time[i] = observation.tstart + (i + 0.5) * bin_duration
+        observation_time[i] = bin_duration
+
+    return evaluation_time, observation_time
```
The current implementation is in a way a bit "brute force" (computing with just a tiny scale). Having an array for observation_time let the possibility later to do much smarter thing with non regular evaluation interval without touching most of the code outside of this function.

mdebony commented on acceptance_modelisation/toolbox.py:
```diff
@@ -82,3 +84,68 @@ def get_unique_wobble_pointings(observations: Observations, max_angular_separati
+    """
+
+    # Compute the scale factor for each mini irf to get a weighted average
+    scale_factor_per_bin = (observation_time / np.sum(observation_time)).to_value(u.dimensionless_unscaled)
```
I maybe missed something but I don't understand why it would be needed as it's already scaled by the total observation time.

mdebony commented on acceptance_modelisation/base_acceptance_map_creator.py:
```diff
@@ -66,7 +68,12 @@ def __init__(self,
             For camera frame transformation the maximum size relative to a pixel a rotation is allowed
         time_resolution : astropy.units.Quantity, optional
-            Time resolution to use for the computation of the rotation of the FoV and cut as function of the zenith bins
+            Time resolution to use for the computation of the rotation of the FoV and cut as function of the zenith binsV
```
Fixed

mdebony commented on acceptance_modelisation/base_acceptance_map_creator.py:
```diff
@@ -663,9 +674,32 @@ def create_acceptance_map_cos_zenith_binned(self,
+                                                              data=data_obs)
+                elif type(dict_binned_model[key_model[0]]) is Background3D:
+                    acceptance_map[obs.obs_id] = Background3D(axes=dict_binned_model[key_model[0]].axes,
+                                                              data=data_obs)
```
Fixed

gabemery replied:
I found this was also an issue in the interpolation case (~100 lines later)

gabemery replied:
Can be fixed here (if merged quickly) or separate PR

mdebony replied:
I've make the modification in the branch improving the interpolation (last PR to make on my side).

mdebony commented on acceptance_modelisation/base_acceptance_map_creator.py:
```diff
@@ -803,4 +851,4 @@ def create_acceptance_map_per_observation(self,
             acceptance_map = self._normalised_model_per_run(observations, acceptance_map)
 
-        return acceptance_map
+        return acceptance_map
```
Fixed

mdebony commented on acceptance_modelisation/base_radial_acceptance_map_creator.py:
```diff
@@ -115,4 +124,4 @@ def create_acceptance_map(self, observations: Observations) -> Background2D:
         acceptance_map = Background2D(axes=[self.energy_axis, self.offset_axis], data=data_background)
 
-        return acceptance_map
+        return acceptance_map
```
Fixed

mdebony commented on acceptance_modelisation/grid3d_acceptance_map_creator.py:
```diff
@@ -222,4 +232,4 @@ def create_acceptance_map(self, observations: Observations) -> Background3D:
                                       fov_alignment=FoVAlignment.ALTAZ)
 
-        return acceptance_map
+        return acceptance_map
```
Fixed

mdebony commented on acceptance_modelisation/radial_acceptance_map_creator.py:
```diff
@@ -111,4 +121,4 @@ def _create_base_computation_map(self, observations: Observation) -> Tuple[WcsND
             livetime += obs.observation_live_time_duration
 
-        return count_map_background, exp_map_background, exp_map_background_total, livetime
+        return count_map_background, exp_map_background, exp_map_background_total, livetime
```
Fixed

mdebony commented on acceptance_modelisation/base_acceptance_map_creator.py:
```diff
@@ -663,9 +674,32 @@ def create_acceptance_map_cos_zenith_binned(self,
+                                                              data=data_obs)
+                elif type(dict_binned_model[key_model[0]]) is Background3D:
+                    acceptance_map[obs.obs_id] = Background3D(axes=dict_binned_model[key_model[0]].axes,
+                                                              data=data_obs)
```
I don't think the marginal performance gain that could be expected (potentially even null depending on how intelligent is the interpreter) would justify the loose in lisibility of the code. Especially as you would still need a condition switch here.

gabemery replied:
I think it would improve the visibility of the code to have a variable called "model_axes" and "model_type".  Instead of always referring to an arbitrary dict_binned_model[key_model[0]].
I don't care about performance here.

Or even a variable BackgroundND that you would use as a class. But this would require to have a dict for kwargs with or without the orientation needed in 3D.

mdebony replied:
The orientation is useless in 2D but supported, so not a big issue.

mdebony replied:
Done as the first proposition.

mdebony commented on acceptance_modelisation/base_acceptance_map_creator.py:
```diff
@@ -66,7 +68,12 @@ def __init__(self,
+            Time resolution to use for the computation of the rotation of the FoV and cut as function of the zenith binsV
+        use_mini_irf_computation : bool, optional
+            If true, during zenith interpolation and binning will compute first mini irf for each part of the run before averaging them.
+            Should improve the accuracy of the model, especially at high zenith angle. Actiate it could singificantly increase computation time.
```
Done

mdebony commented on acceptance_modelisation/base_radial_acceptance_map_creator.py:
```diff
@@ -52,6 +54,11 @@ def __init__(self,
             For camera frame transformation the maximum size relative to a pixel a rotation is allowed
         time_resolution : astropy.units.Quantity, optional
             Time resolution to use for the computation of the rotation of the FoV and cut as function of the zenith bins
+        use_mini_irf_computation : bool, optional
+            If true, during zenith interpolation and binning will compute first mini irf for each part of the run before averaging them.
+            Should improve the accuracy of the model, especially at high zenith angle. Actiate it could singificantly increase computation time.
+        mini_irf_time_resolution : astropy.units.Quantity, optional
+            Time resolution to use for mini irf used for computation of the final background model
```
Done

mdebony commented on acceptance_modelisation/base_acceptance_map_creator.py:
```diff
@@ -66,7 +68,12 @@ def __init__(self,
-            Time resolution to use for the computation of the rotation of the FoV and cut as function of the zenith bins
+            Time resolution to use for the computation of the rotation of the FoV and cut as function of the zenith binsV
+        use_mini_irf_computation : bool, optional
+            If true, during zenith interpolation and binning will compute first mini irf for each part of the run before averaging them.
```
I tried to improved documentation, let me know if it's better.

mdebony commented:
I've addressed the comments on documentation and typo.

For the question of _observation_time_ as an array, it's mainly for more easily improving the functionality later. The current implementation is in a way a bit "brute force" (computing with just a tiny scale). Having an array for observation_time let the possibility later to do much smarter thing with non regular evaluation interval without touching most of the code outside of the time interval function. I think the impact on complexity is marginal.

gabemery commented on acceptance_modelisation/base_acceptance_map_creator.py:
```diff
@@ -10,8 +10,8 @@
 from gammapy.data import Observations, Observation
 from gammapy.datasets import MapDataset
-from gammapy.irf import Background2D, Background3D
-from gammapy.irf.background import BackgroundIRF
```
I would leave the imports possible at higher level (in irf here) as such. The aim woulld be to import BackgroundIRF with gammapy.irf once we drop gammapy 1.1 support

mdebony replied:
I agree, I will try to clean this.

gabemery commented on acceptance_modelisation/grid3d_acceptance_map_creator.py:
```diff
@@ -4,7 +4,8 @@
 import numpy as np
 from gammapy.data import Observations
-from gammapy.irf import Background3D, FoVAlignment
+from gammapy.irf import FoVAlignment
+from gammapy.irf.background import Background3D
```
This is a useless change? No?

gabemery commented on acceptance_modelisation/base_radial_acceptance_map_creator.py:
```diff
@@ -3,7 +3,7 @@
 import numpy as np
 from gammapy.data import Observations
-from gammapy.irf import Background2D
+from gammapy.irf.background import Background2D
```
This is a useless change? No?

gabemery reviewed (commented):
Last commit should no exist in my opinion
